frappe/desk/doctype/number_card/number_card.py

```diff
@@ -109,12 +115,27 @@ def calculate_previous_result(doc, filters):
 	current_date = frappe.utils.now()
 	if doc.stats_time_interval == 'Daily':
 		previous_date = add_to_date(current_date, days=-1)
+		previous_date_from = add_to_date(current_date, days=-2)
 	elif doc.stats_time_interval == 'Weekly':
 		previous_date = add_to_date(current_date, weeks=-1)
+		previous_date_from = add_to_date(current_date, weeks=-2)
 	elif doc.stats_time_interval == 'Monthly':
 		previous_date = add_to_date(current_date, months=-1)
+		previous_date_from = add_to_date(current_date, months=-2)
 	else:
 		previous_date = add_to_date(current_date, years=-1)
+		previous_date_from = add_to_date(current_date, years=-2)
+
+	filters = frappe.parse_json(filters)
+	old_filter = filters
+	filters = []
+	for filter in old_filter:
+		if "Timespan" in filter:
+			cond = str(filter[1])
+			filters.append([doc.document_type, cond, '>', previous_date_from])
+			filters.append([doc.document_type, cond, '<', previous_date])
+		else:
+			filters.append(filter)
```

Comment:
It isn't very clear what you are trying to do here. Can you add some comments or explain in more detail what exactly this piece of code does, and why it is required?

Reply:
The timespan filter has options like This/Last/Next Day/Week/Month.

Let's assume I kept a timespan filter on my number card such that I get the count of documents for a week, ie, Posting Date: Timespan equals This Week.

Now when the system calculates the count of documents created last week(creation date < this week), this filter(posting date = this week) will be present there also, making it look like 0 documents were created.

Reply:
Okay thanks, the issue is clear to me now! I don't think this is the correct way to handle it though, as ideally the filter should be applied while calculating the previous result too.

But since that doesn't make sense either, I think for this case (i.e if a timespan filter is applied), we shouldn't calculate the percentage difference at all.

Reply:
This specific issue arose when a client wanted a dashboard based on weekly data. And he wanted to see the percentage rise from last week as well.

I had to apply timespan filter to all number cards and charts to make the dashboard weekly and had to modify the core to include this.

I do agree that the solution is not exactly the best one. But currently this is the only method I found to have a weekly dashboard.

Maybe we could incorporate an option while creating dashboards to make it weekly/monthly/quarterly/yearly rather than applying timespan filter on each number card and chart.

Reply:
> Maybe we could incorporate an option while creating dashboards to make it weekly/monthly/quarterly/yearly rather than applying timespan filter on each number card and chart.


Yeah I think we can do something like this. For now let's merge the percentage difference fix and work on this in a separate PR?

Reply:
> > Maybe we could incorporate an option while creating dashboards to make it weekly/monthly/quarterly/yearly rather than applying timespan filter on each number card and chart.
> 
> Yeah I think we can do something like this. For now let's merge the percentage difference fix and work on this in a separate PR?

Agreed. I've updated the code.
Thank you :)